seed/validate.go

```diff
@@ -24,62 +24,115 @@ import (
 	"fmt"
 	"path/filepath"
 	"regexp"
+	"sort"
 
-	"github.com/snapcore/snapd/seed/internal"
 	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/timings"
 )
 
+type ValidationError struct {
+	// SystemErrors maps system labels ("" for UC16/18) to their validation errors.
+	SystemErrors map[string][]error
+}
+
+func newValidationError(label string, err error) *ValidationError {
+	return &ValidationError{SystemErrors: map[string][]error{
+		label: {err},
+	}}
+}
+
+func (e *ValidationError) addErr(label string, errs ...error) {
+	if e.SystemErrors == nil {
+		e.SystemErrors = make(map[string][]error)
+	}
+	for _, err := range errs {
+		e.SystemErrors[label] = append(e.SystemErrors[label], err)
+	}
+}
+
+func (e ValidationError) hasErrors() bool {
+	return len(e.SystemErrors) != 0
+}
+
+func (e *ValidationError) Error() string {
+	systems := make([]string, 0, len(e.SystemErrors))
+	for s := range e.SystemErrors {
+		systems = append(systems, s)
+	}
+	sort.Strings(systems)
+	var buf bytes.Buffer
+	first := true
+	for _, s := range systems {
+		if first {
+			if s == "" {
+				fmt.Fprintf(&buf, "cannot validate seed:")
+			} else {
+				fmt.Fprintf(&buf, "cannot validate seed system %q:", s)
+			}
+		} else {
+			fmt.Fprintf(&buf, "\nand seed system %q:", s)
+		}
+		for _, err := range e.SystemErrors[s] {
+			fmt.Fprintf(&buf, "\n - %s", err)
+		}
+	}
+	return buf.String()
+}
+
 // ValidateFromYaml validates the given seed.yaml file and surrounding seed.
 func ValidateFromYaml(seedYamlFile string) error {
-	seed, err := internal.ReadSeedYaml(seedYamlFile)
+	// TODO:UC20: support validating also one or multiple UC20 seed systems
+	// introduce ListSystems ?
+	// What about full empty seed dir?
+	seedDir := filepath.Dir(seedYamlFile)
+
+	seed, err := Open(seedDir, "")
 	if err != nil {
-		return err
+		return newValidationError("", err)
+	}
+
+	if err := seed.LoadAssertions(nil, nil); err != nil {
+		return newValidationError("", err)
 	}
 
-	var errs []error
-	var haveCore, haveSnapd bool
+	tm := timings.New(nil)
+	if err := seed.LoadMeta(tm); err != nil {
```

Comment:
Will this hash all seed snaps? if so I think it might be a bit slow and we need to make sure it is only used if no seeding has happened. Something like https://github.com/snapcore/snapd/compare/master...mvo5:snap-deubg-state-get?expand=1 could help then, i.e. we could add code into the postinst to first check `snap debug state --snap core --get seed.loaded /var/lib/snapd/state.json` != true and only then do the seed validation.

Reply:
yes, it will hash all snaps